polygeist/test/lib/Dialect/Polygeist/TestReachingDefinitionsAnalysis.cpp

```diff
@@ -0,0 +1,96 @@
+//===- TestReachingDefinitionsAnalysis.cpp -------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Analysis/AliasAnalysis.h"
+#include "mlir/Analysis/DataFlow/ConstantPropagationAnalysis.h"
+#include "mlir/Analysis/DataFlow/DeadCodeAnalysis.h"
+#include "mlir/Dialect/Polygeist/Analysis/ReachingDefinitionAnalysis.h"
+#include "mlir/Dialect/SYCL/Analysis/AliasAnalysis.h"
+#include "mlir/Pass/Pass.h"
+#include "llvm/ADT/STLExtras.h"
+
+using namespace mlir;
+using namespace mlir::dataflow;
+
+namespace {
+
+struct TestReachingDefinitionAnalysisPass
+    : public PassWrapper<TestReachingDefinitionAnalysisPass, OperationPass<>> {
+  MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(
+      TestReachingDefinitionAnalysisPass)
+
+  StringRef getArgument() const override { return "test-reaching-definition"; }
+
+  void runOnOperation() override {
+    AliasAnalysis &aliasAnalysis = getAnalysis<mlir::AliasAnalysis>();
+    aliasAnalysis.addAnalysisImplementation(
+        sycl::AliasAnalysis(false /* relaxedAliasing*/));
+
+    DataFlowSolver solver;
+    solver.load<DeadCodeAnalysis>();
+    solver.load<SparseConstantPropagation>();
+    solver.load<polygeist::ReachingDefinitionAnalysis>(aliasAnalysis);
+
+    Operation *op = getOperation();
+    if (failed(solver.initializeAndRun(op)))
+      return signalPassFailure();
+
+    auto printOps = [](std::optional<ArrayRef<Operation *>> ops,
```

Comment:
can it use `ReachingDefinition operator<<` to print, and remove this lambda?

Reply:
Please clarify?

Reply:
I was thinking that instead of 
```
      // Print the reaching definitions for each operand.
      for (auto [index, operand] : llvm::enumerate(op->getOperands())) {
        llvm::errs() << " operand #" << index << "\n";
        std::optional<ArrayRef<Operation *>> mods =
            reachingDef->getModifiers(operand);
        std::optional<ArrayRef<Operation *>> pMods =
            reachingDef->getPotentialModifiers(operand);
        printOps(mods, " - mods: ");
        printOps(pMods, " - pMods: ");
      }
```
we could do
```
llvm::errs() << *reachingDef;
```
I guess we cannot, as `operator<<(raw_ostream &, const ReachingDefinition &)` doesn't have:
```
            if (auto tagName = op->getAttrOfType<StringAttr>("tag_name"))
              llvm::errs() << tagName.getValue();
```
Do you think is worth having that there?

Reply:
The "tag_name" has been added to the lit tests exclusively to make writing the lit test easier (I got the idea from other dataflow lit tests in MLIR). Therefore the tag_name is not germane to the actual `ReachingDefinition` class so we shouldn't print it in that class `operator<<`